apps/dcos_overlay/src/dcos_overlay_lashup_kv_listener.erl

```diff
@@ -1,218 +1,127 @@
-    StateData#data{config = [], num_res = Res + 1}; %% clean cached config
-maybe_update_state(StateData = #data{num_res = Res}) ->
-    StateData#data{num_res = Res + 1}.
-
-next_state_transition(StateData = #data{num_req = Req, num_res = Res}) when Req == Res ->
-    ReapplyTimeout = application:get_env(dcos_overlay, reapply_timeout, ?REAPPLY_TIMEOUT),
-    {next_state, batching, StateData, {timeout, ReapplyTimeout, do_reapply}};
-next_state_transition(StateData) ->
-    {keep_state, StateData}.
-
--spec(fetch_lashup_config() -> config()).
-fetch_lashup_config() ->
-    MatchSpec = mk_key_matchspec(),
-    OverlayKeys = lashup_kv:keys(MatchSpec),
-    lists:foldl(
-        fun(OverlayKey, Acc) ->
-            KeyValue = lashup_kv:value(OverlayKey),
-            orddict:append_list(OverlayKey, KeyValue, Acc)
-        end,
-        orddict:new(), OverlayKeys).
-
-next_state_and_actions(Config, StateData0) ->
-    Actions = create_actions(Config),
-    {ok, Timer} = timer:kill_after(?KILL_TIMER),
-    StateData1 = StateData0#data{kill_timer = Timer, num_req = length(Actions), num_res = 0},
-    {StateData1, Actions}.
-
-create_actions(Config) ->
-    create_actions(Config, []).
-
-create_actions([], Acc) ->
-    lists:reverse(Acc);
-create_actions([Config|Configs], Acc) ->
-    Action = create_action(Config),
-    create_actions(Configs, [Action|Acc]).
-
-create_action({OverlayKey, KeyValue}) ->
-    EventContent = #{key => OverlayKey, value => KeyValue},
-    {next_event, internal, EventContent}.
+    {ok, {}, {continue, {}}}.
+
+handle_continue({}, {}) ->
+    ok = wait_for_vtep(),
+    MatchSpec = ets:fun2ms(fun({?KEY('_')}) -> true end),
+    {ok, Ref} = lashup_kv:subscribe(MatchSpec),
+    RRef = start_reconcile_timer(),
+    {noreply, #state{ref=Ref, reconcile_ref=RRef}}.
+
+handle_call(_Request, _From, State) ->
+    {noreply, State}.
+
+handle_cast(_Request, State) ->
+    {noreply, State}.
+
+handle_info({lashup_kv_event, Ref, Key},
+        #state{ref=Ref, config=Config}=State) ->
+    ok = lashup_kv:flush(Ref, Key),
+    Value = lashup_kv:value(Key),
+    {Subnet, Delta, Config0} = update_config(Key, Value, Config),
+    ok = apply_configuration(#{Subnet => Delta}),
+    {noreply, State#state{config=Config0}};
+handle_info({timeout, RRef0, reconcile},
+        #state{config=Config, reconcile_ref=RRef0}=State) ->
+    ok = apply_configuration(Config),
+    RRef = start_reconcile_timer(),
+    {noreply, State#state{reconcile_ref=RRef}};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
 
+-spec(wait_for_vtep() -> ok).
+wait_for_vtep() ->
+    % listener must wait until vtep is configured.
+    % dcos_overlay_poller polls local mesos agent module and
+    % configures vtep interfaces.
+    try
+        _Overlays = dcos_overlay_poller:overlays(),
```

Comment:
what if `overlays` returns with an empty list? We would like to wait until we have a non-empty list of overlays.

Reply:
Oops. Good catch! I didn't notice this case. Fixed.